<a href="https://colab.research.google.com/github/RodolfoFigueroa/madi2022-1/blob/main/Inducci%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Antes de iniciar siempre es importante agregar al inicio las librerías que vamos a usar, y compilar la celda correspondiente.

In [ ]:
import numpy as np
import time 

Es importante considerar que las celdas de código tienen "memoria", por ejemplo, en la siguiente celda definiremos el valor de una variable, y en una celda posterior la vamos a imprimir, si pidiéramos imprimirla sin antes haberle asignado un valor (es decir sin haber compilado la celda previa) estaríamos en problemas.

In [ ]:
a = 70

In [ ]:
b = 10

In [ ]:
print(a)
print(b)

70
10


En esta sesión veremos algunos ejemplos de inducción y cómo este principio nos puede ayudar a ser más eficientes al resolver un problema.

**Ejemplo 1.**  Encuentra el valor de $$\Big\lfloor \frac{1}{2}\Big \rfloor + \Big\lfloor \frac{2}{2}\Big \rfloor + \cdots + \Big\lfloor \frac{10^9}{2}\Big \rfloor$$

*Paso 1*. Hagamos algunos casos pequeños para encontrar algún patrón.

In [ ]:
s = 0
for i in range (1, 10):
  s = s + np.floor(i/2)
  print(s, " con n = ", i)

0.0  con n =  1
1.0  con n =  2
2.0  con n =  3
4.0  con n =  4
6.0  con n =  5
9.0  con n =  6
12.0  con n =  7
16.0  con n =  8
20.0  con n =  9


*Paso 2.* ¿Alguna conjetura? Probaremos por inducción que $$\Big\lfloor \frac{1}{2}\Big \rfloor + \Big\lfloor \frac{2}{2}\Big \rfloor + \cdots + \Big\lfloor \frac{n}{2}\Big \rfloor = \Big \lfloor \frac{n}{2} \Big \rfloor \Big \lfloor \frac{n+1}{2}\Big \rfloor$$
 - Base de inducción. Con los casos que hemos checado anteriormente es más que suficiente.
 - Hipótesis de inducción. La fórmula que queremos probar.
 - Paso inductivo. Supongamos que nuestra fórmula es cierta para $k$. Para $k+1$ se sigue fácilmente del hecho $$ \Big \lfloor \frac{k}{2} \Big \rfloor \Big \lfloor \frac{k+1}{2}\Big \rfloor + \Big \lfloor \frac{k+1}{2}\Big \rfloor = \Big \lfloor \frac{k+1}{2}\Big \rfloor (\Big \lfloor \frac{k}{2}\Big \rfloor + 1) = \Big \lfloor \frac{k+1}{2} \Big \rfloor \Big \lfloor \frac{k+2}{2}\Big \rfloor$$



*Paso 3.* Concluimos calculando el valor que se nos ha pedido.

In [ ]:
start = time.time()
print(np.floor(1e9/2)*np.floor((1e9+1)/2))
print(time.time() - start)

2.5e+17
0.0008492469787597656


Comparemos el tiempo de ejecución con el tiempo que toma hacer la operación de forma directa. (Haremos la suma hasta $10^6$, pues tarda mucho para $10^9$).

In [ ]:
start = time.time()
s = 0
for i in range (1, 1000001):
  s += np.floor(i/2)

print(s)
print(time.time() - start)

250000000000.0
1.144911289215088


Esto nos muestra que fue de gran ayuda encontrar una fórmula cerrada, para reducir bastante el tiempo de ejecución (pasa de ser lineal a ser constante).

**Ejemplo 2.** Sea $f : \mathbb{N} \rightarrow \mathbb{R}$ una función tal que $f(1) = 1001$ y $f(1) + f(2) + \cdots + f(n) = n^2 f(n)$ para todo entero $n > 1$. Encuentra el valor de $f(10007)$.

*Paso 1.* Primero simplifiquemos nuestras cuentas. Para eso, notemos que si $g(m) = \frac{f(m)}{1001}$, se tiene que $g(1) + g(2) + \cdots + g(n) = n^2g(n)$ para todo $n > 1$, entonces $g(n)$ depende de los primeros $n-1$ valores: $$(n^2 -1)g(n) = g(1) + g(2) + \cdots g(n-1)$$, y como $((n-1)^2 - 1)g(n-1) = g(1) + g(2) + \cdots g(n-2)$, sustituyendo se tiene que $$(n^2 -1)g(n) = ((n-1)^2 - 1)g(n-1) + g(n-1) = (n-1)^2g(n-1)$$.De esto obtenemos que $g(n) = \frac{n-1}{n + 1}g(n-1)$. Ahora hagamos los primeros casos.

In [ ]:
g = 1
for i in range (2, 10):
  g = g*(i-1)/(i+1)
  print(g)

0.3333333333333333
0.16666666666666666
0.1
0.06666666666666667
0.047619047619047616
0.03571428571428571
0.027777777777777776
0.02222222222222222


*Paso 2*. ¿Conjetura? Si uno no se sabe la forma decimal de muchas fracciones de memoria conjeturar algo de aquí puede ser difícil. ¿Qué pasa si calculamos $h = 1/g$? Ahora tenemos que $h(n) = \frac{n+1}{n-1}h(n-1)$

In [ ]:
h = 1
for i in range (2, 10):
  h = h*(i+1)/(i-1)
  print("h(", i, ")= ", h)

h( 2 )=  3.0
h( 3 )=  6.0
h( 4 )=  10.0
h( 5 )=  15.0
h( 6 )=  21.0
h( 7 )=  28.0
h( 8 )=  36.0
h( 9 )=  45.0


*Paso 3.* Podemos conjeturar que $h(n) = \frac{n(n+1)}{2}$, y es fácil probar por inducción que esta fórmula es correcta, usando que $h(n) = \frac{n+1}{n-1}h(n-1)$ para $n > 1$.

Finalmente, recordando que buscamos $f(10007)$, calculamos usando la fórmula que encontramos:

In [ ]:
print(2/(10007*10008*1001))

1.9950083682449443e-11


**Ejemplo 3.** Definimos $F_n = 2^{2^n} + 1$, el $n$ -ésimo número de Fermat, para $n = 0, 1, \dots$. Muestra que cualesquiera dos números de Fermat son primos relativos ($a$ y $b$ son primos relativos si su máximo común divisor es $1$).


*Paso 1.* Casos pequeños. No podemos hacer muchos casos porque $2^{2^n}$ es muy grande, veamos qué pasa con los primeros $6$ números de Fermat.

In [ ]:
print("Los primeros seis números de Fermat son:")
F = []
for i in range (0, 6):
  F.append(pow(2, pow(2, i)) + 1)
  print(F[i])

Los primeros seis números de Fermat son:
3
5
17
257
65537
4294967297


*Paso 2.* A primera vista no hay mucho de dónde agarrarnos, intentando buscar un patrón notamos que $3 = 5 -2$,  $3 \times 5 = 17 - 2$ y que $3 \times 5 \times 17 = 255 = 257 - 2$. ¿Esto nos sirve? Sí, pues es fácil ver que $F_n$ siempre es impar, por lo que si el producto de todos los números de Fermat anteriores nos da $F_n - 2$, se tiene que en particular $(F_k, F_n) = 1$ para todo $k < n$, de donde se concluye lo deseado. Aquí tenemos una primera conjetura, veamos si se cumple para el resto de los números calculados.

In [ ]:
P = F[0]
for i in range (0, 5):
  if (F[i+1] - P == 2):
    print("Se cumple")
  else:
    print("No se cumple :c")
  if(i+1 < 5) : P *= F[i+1]

Se cumple
Se cumple
Se cumple
Se cumple
Se cumple


*Paso 3*. Nuestros casos parecen indicar que $F_1 \cdot F_2 \cdots F_n = F_{n+1}-2$. Probemos esto por inducción.

- Caso base. Los casos que ya vimos anteriormente.
- Hipótesis de inducción. Para $n = k$ se cumple que $F_1 \cdot F_2 \cdot \dots \cdot F_k = F_{k+1}-2$. 
- Paso inductivo. Usando la hipótesis de inducción, tenemos que: $$F_1 \cdot F_2 \cdots F_{k} \cdot F_{k+1} = (F_{k+1} - 2)F_{k+1}$$, recordando que $F_{n} = 2^{2^n}+1$, tenemos que 
\begin{align}
(F_{k+1} - 2)F_{k+1} &= (2^{2^{k+1}} - 1)(2^{2^{k+1}} + 1) \\
&= 2^{2^{k+2}} - 1 \\
&= F_{k+2} - 2
\end{align}

De donde concluimos el paso inductivo, y por lo tanto que se cumple $F_1 \cdot F_2 \cdots F_n = F_{n+1}-2$ para todo entero positivo $n$. Con esto hemos probado que cualesquiera dos números de Fermat distintos son primos relativos, como queríamos ver.

NOTA : Notemos que esto nos permite además concluir que hay una infinidad de números primos.

**Ejemplo 4.** Considera el conjunto $A = \{1, 2, \dots, 1000 \}$. Para cada subconjunto no vacío $B$ de $A$ definimos $f(B)$ como el producto de los cuadrados de los elementos de $B$, por ejemplo, si $B = \{1, 5\}$, $f(B) = 1^2 \cdot 5^2 = 25$. Sea $S$ el conjunto de todos los subconjuntos no vacíos de $A$ tales que no contienen parejas de enteros consecutivos, por ejemplo $\{1, 4\} \in S$, pero $\{2, 3, 7\} \notin S$. Determina el valor de $$F(A) = \sum_{B \in S}f(B)$$.

*Paso 1.* Se siente que el valor de la suma pedida para el conjunto de los primeros $n$ enteros positivos va a ir creciendo muy rápido conforme $n$ crece, entonces calculemos los primeros valores, pero no muchos para evitar que nuestras computadoras exploten. ¿Cómo escribir un código para que calcule los primeros casos sin hacer casito por casito? Hint: Ver que $F([k]) = F([k-1]) + F([k-2])k^2 + k^2$, donde $[k] = \{1, 2, \dots, k\}$.

In [ ]:
s = [0, 0]
for i in range (1, 6):
  j = i%2
  s[j] = s[(j+1)%2] + s[j]*i*i + i*i 
  print("Para n = ", i, " da ", s[j])

Para n =  1  da  1
Para n =  2  da  5
Para n =  3  da  23
Para n =  4  da  119
Para n =  5  da  719


*Paso 2.* Notemos que al sumar $1$ a cada valor obtenemos un factorial. Probemos por inducción que en efecto esto sucede, y de hecho la prueba va a resultar muy fácil gracias a la lógica que usamos para el código de los primeros casos. Denotaremos $[n] = \{1, 2, \dots, n\}$.

- Caso base. Los casos que ya vimos.
- Hipótesis de inducción. $F([k]) = (k+1)! - 1$ para todo entero positivo $k \leq n$.
- Paso inductivo. Supongamos que $F([k]) = (k+1)! - 1$ para todo entero positivo $k \leq n$. Como vimos para hacer el código de los primeros casos, $$F([n+1]) = F([n]) + (n+1)^2F([n-1]) + (n+1)^2$$ sustituyendo, tenemos que 
\begin{align}
F([n+1]) &= ((n+1)! - 1) + ((n)! - 1)(n+1)^2 + (n+1)^2 \\
&= (n+1)! + n!(n+1)^2 - 1 \\
&= (n+1)!(1 + n+1) - 1 \\
&= (n+2)! - 1
\end{align}

Con lo que concluimos que $F([n+1]) = (n+2)! - 1$.

Entonces la respuesta a nuestro problema inicial es $1001! - 1$, que es muy muy grande.

**Ejercicios.** Aquí va una pequeña lista de ejercicios para que practiquen su uso de la técnica de inducción matemática.

1.   Encuentra el valor de $1^3 + 3^3 + 5^3 + \cdots + 100005^3$.
2.   Considera todos los subconjuntos no vacíos de $\{1, 2, \dots, 2048\}$. Para cada uno de estos subconjuntos, considera el recíproco del producto de sus elementos. Calcula la suma de estos números. 
3.   (Reto, no obligatorio) Considera el conjunto $\textit{F}$ de todas las funciones inyectivas $f : \mathbb{Z}^{+} \rightarrow \mathbb{Z}^{+}$ que satisfacen $f(2x) + f(x)f(y) = f(xy) + 2f(x)$ para cualesquiera $x, y \in \mathbb{N}$. Determina el valor de $min_{f \in \textit{F}} \{f(2012)\}$.

*Ejercicio 1.* Encuentra el valor de $1^3 + 3^3 + 5^3 + \cdots + 100005^3$. 

- Intenta encontrar primero una fórmula para $1^3 + 2^3 + \cdots + n^3$, crea un código para checar casos pequeños en la siguiente celda.

- ¿Alguna conjetura? Pruébala usando inducción. ¿Cómo pasar de esta suma a sumar únicamente los cubos de los impares? Quizás sea conveniente calcular la respectiva suma para los números pares, es decir, $2^3 + 4^3 + \cdots + (2k)^3$.
- Una vez que tengas una fórmula para la suma buscada, escribe un código que imprima el valor buscado y agrega tu demostración de la fórmula.

(Aquí va la demostración)

In [ ]:
print("El valor de la suma dada es: ",   )

El valor de la suma dada es: 


*Ejercicio 2.* Recuerda lo que se hizo en el ejemplo $4$.

- Escribe un código que te permita calcular los primeros casos en la siguiente celda.

- Prueba tu conjetura usando inducción. 
- Escribe un código que imprima el valor que se pide.

(Aquí va la prueba de tu conjetura)

In [ ]:
print("La suma de los números indicados es :", )

La suma de los números indicados es :
